### PR TITLE
Restrict CASA Table Access to a Single Thread per Table

### PR DESCRIPTION
**Pull request by sjperkins**
CASA Tables are not threadsafe and should only be accessed by the thread that opened them. Restrict access to a single thread.

- [x] Tests added / passed

  ```bash
  $ py.test -v -s xarrayms/tests
  ```

  If the pep8 tests fail, the quickest way to correct
  this is to run `autopep8` and then `flake8` and
  `pycodestyle` to fix the remaining issues.

  ```
  $ pip install -U autopep8 flake8 pycodestyle
  $ autopep8 -r -i xarrayms
  $ flake8 xarrayms
  $ pycodestyle xarrayms
  ```

- [x] Fully documented, including `HISTORY.rst` for all changes
      and one of the `docs/*-api.rst` files for new API

  To build the docs locally:

  ```
  pip install -r requirements.readthedocs.txt
  cd docs
  READTHEDOCS=True make html
  ```

### CODE REVIEW

**Comment by sjperkins** [comment]
@bennahugo Can you please try this branch?

**Comment by bennahugo** [comment]
Nope still breaks:
```
[###############################         ] | 79% Completed |  1hr 26min 22.0s^[[B^[[B^[[B
Unexpected error. Dropping you into pdb for a post-mortem.
Traceback (most recent call last):
  File "/home/hugo/venv/bin/tricolour", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/home/hugo/tricolour/tricolour/scripts/tricolour", line 5, in <module>
    tricolour.main()
  File "/home/hugo/tricolour/tricolour/__init__.py", line 495, in main
    dask.compute(write_computes)
  File "/home/hugo/venv/local/lib/python2.7/site-packages/dask/base.py", line 398, in compute
    results = schedule(dsk, keys, **kwargs)
  File "/home/hugo/venv/local/lib/python2.7/site-packages/dask/threaded.py", line 76, in get
    pack_exception=pack_exception, **kwargs)
  File "/home/hugo/venv/local/lib/python2.7/site-packages/dask/local.py", line 462, in get_async
    raise_exception(exc, tb)
  File "/home/hugo/venv/local/lib/python2.7/site-packages/dask/local.py", line 230, in execute_task
    result = _execute_task(task, data)
  File "/home/hugo/venv/local/lib/python2.7/site-packages/dask/core.py", line 119, in _execute_task
    return func(*args2)
  File "/home/hugo/venv/local/lib/python2.7/site-packages/xarrayms/xarray_ms.py", line 245, in _chunk_putcols_np
    startrow=rs, nrow=rl)   
  File "/home/hugo/venv/local/lib/python2.7/site-packages/xarrayms/table_proxy.py", line 71, in __call__
    return getattr(table, fn)(*args, **kwargs)
  File "/home/hugo/venv/local/lib/python2.7/site-packages/casacore/tables/table.py", line 1157, in putcol
    self._putcol(columnname, startrow, nrow, rowincr, value)
RuntimeError: ColumnSet::doLock: table /scratch/bhugo/G330_2/msdir/1524852817.1GC.ms should be locked when using UserLocking
> /home/hugo/venv/local/lib/python2.7/site-packages/casacore/tables/table.py(1157)putcol()
-> self._putcol(columnname, startrow, nrow, rowincr, value)
```

File: xarrayms/table_cache.py [hunk 1]
```diff
@@ -80,8 +72,6 @@ def acquire(self, locktype):
         else:
             raise ValueError("Invalid lock type %d" % locktype)
 
-        self.refcount += 1
-
     def release(self, locktype):
         if locktype == READLOCK:
             self.readlocks -= 1
```
Comment (bennahugo) [inline xarrayms/table_cache.py]: Potential race check then update race condition here?

Comment (sjperkins) [reply xarrayms/table_cache.py]: No, the refcount isn't used for anything interesting at the moment

**Comment by sjperkins** [comment]
@bennahugo Can you try the commit I've just pushed up?

**Comment by bennahugo** [comment]
it ran through 2 times, however this isn't a guarantee the race condition is fixed though. As discussed yesterday the safest thing to do is to implement the producer consumer pattern here with one io thread spinning off data onto disk. It is very suboptimal, but I don't think it can be helped with the current way casacore does IO.

See:
https://github.com/bennahugo/politsiyakat/blob/master/politsiyakat/data/data_provider.py#L197
for an example.

**Pull request merged**